File: spark-event-ui/src/app/model/trackers/metrics/MetricDescriptionRegistry.ts
import {registerStdMetricDescriptions} from './MetricDescriptions';
import {EnumMetricType, GenericMetricDescription, MetricDescription} from './MetricDescription';

/**
 *
 */
export class MetricDescriptionRegistry {
  static readonly INSTANCE = new MetricDescriptionRegistry();

  readonly registry: Map<string, MetricDescription> = new Map<string, MetricDescription>();

  constructor() {
    registerStdMetricDescriptions(this);
  }

  byName(name: string) {
    return this.registry.get(name);
  }

  register(type: MetricDescription): void {
    this.registry.set(type.name, type);
  }

  registers(types: MetricDescription[]): void {
    types.forEach(t => this.register(t));
  }

  resolve(name: string, metricType: string): MetricDescription {
    let found = this.byName(name);
    if (found) {
      return found;
    } else {
      // NOT found! ... register a generic one
      // console.log("MetricDescriptionRegistry: registering generic metric description for name='" + name + "', type='" + metricType + "'");

      console.log("export class " + name + "MetricDescription extends MetricDescription {\n" +
                  "  constructor() {\n" +
                  "    super('" + name + "', '" + metricType + "', 'unit?');\n" +
                  "  }\n" +
                  "}\n");

      const generic = new GenericMetricDescription(name, metricType as EnumMetricType);
      this.register(generic);
      return generic;
    }
  }
}

File: spark-event-gateway-srv/src/main/java/fr/an/spark/gateway/eventTrackers/sql/metrics/MetricDescriptionRegistry.java
package fr.an.spark.gateway.eventTrackers.sql.metrics;

import fr.an.spark.gateway.eventTrackers.sql.metrics.MetricDescription.EnumMetricType;
import fr.an.spark.gateway.eventTrackers.sql.metrics.MetricDescription.GenericMetricDescription;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MetricDescriptionRegistry {
    public static final MetricDescriptionRegistry INSTANCE = new MetricDescriptionRegistry();

    private final Map<String, MetricDescription> registry = new HashMap<>();

    private MetricDescriptionRegistry() {
        MetricDescriptions.registerStdMetricDescriptions(this);
    }

    public MetricDescription byName(String name) {
        return registry.get(name);
    }

    public void register(MetricDescription type) {
        registry.put(type.getName(), type);
    }

    public void registers(List<MetricDescription> types) {
        for (MetricDescription t : types) {
            register(t);
        }
    }

    public MetricDescription resolve(String name, String metricType) {
        MetricDescription found = byName(name);
        if (found != null) {
            return found;
        } else {
            System.out.println("export class " + name + "MetricDescription extends MetricDescription {\n" +
                    "  constructor() {\n" +
                    "    super('" + name + "', '" + metricType + "', 'unit?');\n" +
                    "  }\n" +
                    "}\n");
            GenericMetricDescription generic = new GenericMetricDescription(name, EnumMetricType.valueOf(metricType));
            register(generic);
            return generic;
        }
    }
}
